Introduce special starting state in ActorAdapter for guardian actor #24279

jrudolph opened the pull request:
An experimental alternative for @johanandren's delayed guardian init PR, #24302.

The idea is to create a special delayed start phase in ActorAdapter so we can defer the need for actual adapter creation until after the untyped ActorSystem has been created.

Just the diff to #24302: https://github.com/johanandren/akka/compare/wip-24279-typed-guardian-woes-johanandren...jrudolph:jr/review-24302.

johanandren committed:
=typ #24279 Add tests for ActorSystem startup / shutdown scenarios

jrudolph committed:
=typ allow creation of ActorSystemAdapters only after ActorSystem is fully initialized

This will allow to uncover initialization order bugs easier.

jrudolph committed:
=typ #24279 introduce special startup state for typed guardian actor

This allows the untyped actorsystem to be initialized completely before starting up
the typed guardian code.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala:
```diff
@@ -14,19 +14,32 @@ import akka.util.OptionVal
+  def receive = running
+
+  def waitingForStart: Receive = {
+    case ActorAdapter.Start ⇒ start()
```
ok, so stashing is not needed because no message should be sent to it before started. The ActorRef (ActorSystem) has not been returned yet. The only case would be if some crazy (invalid) extension would send to the system when loaded eagerly. I'm fine with not supporting that.

patriknw reviewed (commented):
looks like a good approach

(some cleanup is obviously missing)

johanandren commented:
Cool, seems reasonable, there was quite a bit of test failures though, are those related?

jrudolph commented:
Quite likely related ;)

johanandren commented:
Refs #24279

jrudolph commented:
@johanandren I remixed your commits from #24302 and added my fix. Do we need the `Behavior` change? What does it do? Any other comments about what to keep and to remove?

In the final version I'll just remove the `delayedStart` fix, right?

patriknw commented:
DelayedStart can be removed, as far as I know

jrudolph commented:
Removed things which didn't seem essential.

jrudolph commented on akka-actor-typed/src/main/scala/akka/actor/typed/ActorSystem.scala:
```diff
@@ -211,15 +212,16 @@ object ActorSystem {
 
-    val adapter: ActorSystemAdapter.AdapterExtension = ActorSystemAdapter.AdapterExtension(untyped)
-    adapter.adapter
+    untyped.guardian ! ActorAdapter.Start
```
I wonder how sure we can be that `Start` will be the first message that arrives at the guardian? It seems we would guarantee that only from the thread that created the `ActorSystem`.

jrudolph commented:
* Fixed a potential race where the `Start` message may be too late. 
 * Moved processing from `ActorAdapter` into specialized `GuardianActorAdapter` to keep the regular code clean
 * Fixed ActorSystem shutdown before `Start` message has been processedd

jrudolph commented:
If the newest changes in https://github.com/akka/akka/pull/24406/commits/f6dac91a76ec6ad58a7805eb99596c151a2daf46 would be reviewed I think this can then be merged.

patriknw commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/adapter/AdapterSpec.scala:
```diff
@@ -157,6 +157,37 @@ class AdapterSpec extends AkkaSpec {
+          system = ActorSystem.create(Behaviors.deferred[NotUsed](_ ⇒ Behavior.stopped[NotUsed]), "AdapterSpec-stopping-guardian")
+        } finally {
+          if (system != null) {
+            Await.result(system.terminate(), remainingOrDefault)
```
I prefer that we use the `shutdown` method in testkit, since it gives better reporting if it fails, and we have the timeout in a single place

jrudolph replied:
I see, good to know.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorAdapter.scala:
```diff
@@ -135,6 +146,57 @@ import akka.util.OptionVal
+ * That will allow to defer typed processing until the untyped ActorSystem has completely started up.
+ */
+@InternalApi
+private[typed] class GuardianActorAdapter[T](_initialBehavior: Behavior[T]) extends ActorAdapter[T](_initialBehavior) {
```
👍

patriknw reviewed (approved):
LGTM

jrudolph commented:
Final fixes.

jrudolph commented:
Fixed #24279.
